faunadb-java-dsl/src/main/java/com/faunadb/client/types/Decoder.java

```diff
@@ -0,0 +1,253 @@
+      return toMap(value, (MapType) dstType);
+
+    return toObject(value, dstType);
+  }
+
+  private static Result<Enum> toEnum(Value value, Class<?> enumClass) {
+    final Map<String, Enum> enumMap = Enums.getDecodingMap((Class<Enum>) enumClass);
+
+    return value.to(Codec.STRING).map(new Function<String, Enum>() {
+      @Override
+      public Enum apply(String input) {
+        return enumMap.get(input);
+      }
+    });
+  }
+
+  private static Result<Map<String, Object>> toMap(Value value, MapType dstType) {
+    try {
+      ImmutableMap<String, Value> values = ((ObjectV) value).values;
+
+      SimpleType valueType = dstType.getValueType();
+
+      Map<String, Object> map = (Map<String, Object>) getConcreteClass(dstType.getRawClass()).newInstance();
+
+      for (Map.Entry<String, Value> entry : values.entrySet()) {
+        map.put(entry.getKey(), decodeImpl(entry.getValue(), valueType).get());
+      }
+
+      return Result.success(map);
+    } catch (Exception ex) {
+      return Result.fail(format("Could not instantiate map of type %s", dstType), ex);
+    }
+  }
+
+  private static Result<Collection<Object>> toCollection(Value value, CollectionType dstType) {
+    try {
+      ImmutableList<Value> values = ((ArrayV) value).values;
+
+      SimpleType elementType = dstType.getElementType();
+
+      Collection<Object> collection = (Collection<Object>) getConcreteClass(dstType.getRawClass()).newInstance();
+
+      for (Value v : values) {
+        collection.add(decodeImpl(v, elementType).get());
+      }
+
+      return Result.success(collection);
+    } catch (Exception ex) {
+      return Result.fail(format("Could not instantiate collection of type %s", dstType), ex);
+    }
+  }
+
+  private static Class<?> getConcreteClass(Class<?> rawClass) {
+    if ((rawClass.getModifiers() & Modifier.ABSTRACT) == 0 && !rawClass.isInterface())
+      return rawClass;
+
+    if (List.class.isAssignableFrom(rawClass))
+      return ArrayList.class;
+
+    if (Set.class.isAssignableFrom(rawClass))
+      return HashSet.class;
+
+    if (Map.class.isAssignableFrom(rawClass))
+      return HashMap.class;
+
+    throw new FaunaException(format("Abstract class not supported: $%s", rawClass));
+  }
+
+  private static Result<Object> toArray(Value value, Class<?> dstType) {
+    Class<?> componentType = dstType.getComponentType();
+
+    if (!(value instanceof ArrayV))
+      return Result.fail(format("Cannot convert %s to an array of type %s", value, componentType));
+
+    ImmutableList<Value> values = ((ArrayV) value).values;
+    int length = values.size();
+
+    Object array = Array.newInstance(componentType, length);
+
+    for (int i = 0; i < length; i++) {
+      Array.set(array, i, decodeImpl(values.get(i), Types.of(componentType)).get());
+    }
+
+    return Result.success(array);
+  }
+
+  private static Result<Value> toValue(Value value, Class<?> dstType) {
+    if (value == null)
+      return Result.<Value>success(NullV.NULL);
```

Comment:
this is for cases when we are decoding from `Value` to `Value` and we have a null instance, for example
```java
class Foo {
  @FaunaField
  Value bar;
}

decode(new ObjectV(mapOf("bar", (Value) null)), Foo.class)
```
consider `mapOf` returns a map that allows null as values. It will return a instance of `Foo` with attribute `bar` populated with `NullV.NULL`

Or more simple:
```java
encode(null, Value.class)
```
it will return `NullV.NULL`